src/librustc_mir/interpret/intern.rs

```diff
@@ -0,0 +1,219 @@
+                    None => self.ref_tracking.track(mplace, || ()),
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+pub fn intern_const_alloc_recursive(
+    ecx: &mut CompileTimeEvalContext<'a, 'mir, 'tcx>,
+    def_id: DefId,
+    ret: MPlaceTy<'tcx>,
+    param_env: ty::ParamEnv<'tcx>,
+) -> EvalResult<'tcx> {
+    let tcx = ecx.tcx;
+    let frozen = ret.layout.ty.is_freeze(tcx.tcx, param_env, tcx.span);
+    let intern_mode = match tcx.is_static(def_id) {
+        Some(hir::Mutability::MutMutable) => InternMode::StaticMut,
+        Some(hir::Mutability::MutImmutable) => InternMode::Static { frozen },
+        None => InternMode::Const { frozen },
+    };
+
+    // type based interning
+
+    let mut ref_tracking = RefTracking::new(ret);
+    let mut leftover_allocations = FxHashSet::default();
+
+    // start with the outer most allocation
+    InternVisitor {
+        ref_tracking: &mut ref_tracking,
+        ecx,
+        mode: intern_mode,
+        leftover_relocations: &mut leftover_allocations,
+    }.intern(ret.ptr.to_ptr()?)?;
+
+    while let Some((mplace, ())) = ref_tracking.todo.pop() {
+        let interned = InternVisitor {
+            ref_tracking: &mut ref_tracking,
+            ecx,
+            mode: intern_mode,
+            leftover_relocations: &mut leftover_allocations,
+        }.visit_value(mplace);
+        if let Err(error) = interned {
+            if let EvalErrorKind::ValidationFailure(_) = error.kind {
+                // FIXME(oli-obk): maybe only create one error with many notes about the kind of
+                // error that occurred
+                let err = crate::const_eval::error_to_const_error(&ecx, error);
+                match err.struct_error(ecx.tcx, "it is undefined behavior to use this value") {
+                    Ok(mut diag) => {
+                        diag.note("The rules on what exactly is undefined behavior aren't clear, \
+                            so this check might be overzealous. Please open an issue on the rust \
+                            compiler repository if you believe it should not be considered \
+                            undefined behavior",
+                        );
+                        diag.emit();
+                    }
+                    Err(ErrorHandled::TooGeneric) |
+                    Err(ErrorHandled::Reported) => {},
+                }
+            }
+        }
+    }
+
+    // Intern the rest of the allocations as mutable. These might be inside unions, padding, raw
+    // pointers, ... So we can't intern them according to their type rules
+
+    let mut todo: Vec<_> = leftover_allocations.iter().cloned().collect();
+    while let Some(alloc_id) = todo.pop() {
+        if let Some((_, alloc)) = ecx.memory_mut().alloc_map.remove(&alloc_id) {
+            let alloc = tcx.intern_const_alloc(alloc);
+            tcx.alloc_map.lock().set_alloc_id_memory(alloc_id, alloc);
```

Comment:
I kind of expected this to call `intern` instead of open-coding the actual interning again. That might not be reasonable, but in that case please add a comment.